README.md
[![Build Status](https://travis-ci.org/sappelhoff/remedian.svg?branch=master)](https://travis-ci.org/sappelhoff/remedian)
[![codecov](https://codecov.io/gh/sappelhoff/remedian/branch/master/graph/badge.svg)](https://codecov.io/gh/sappelhoff/remedian)
[![Documentation Status](https://readthedocs.org/projects/remedian/badge/?version=latest)](http://remedian.readthedocs.io/en/latest/?badge=latest)
[![PyPI version](https://badge.fury.io/py/remedian.svg)](https://badge.fury.io/py/remedian)

# remedian
The  Remedian:  A  Robust  Averaging  Method  for  Large  Data  Sets - Python
implementation

This algorithm is used to approximate the median of several data chunks if
these data chunks cannot (or should not) be loaded into memory at once.

Given a data chunk of size `obs_size`, and `t` data chunks overall, the
Remedian class sets up a number `k_arrs` of arrays of length `n_obs`.

The median of the `t` data chunks of size `obs_size` is then approximated
as follows: One data chunk after another is fed into the `n_obs` positions
of the first array. When the first array is full, its median is calculated
and stored in the first position of the second array. After this, the first
array is re-used to fill the second position of the second array, etc.
When the second array is full, the median of its values is stored in the
first position of the third array, and so on.

The final "Remedian" is the median of the last array, after all `t` data
chunks have been fed into the object.

# Installation

`pip install remedian`

The dependencies should be installed automatically by pip.

# Installation of most recent version

1. activate your python environment
2. `git clone https://www.github.com/sappelhoff/remedian`
3. `cd remedian`
5. `pip install -e .`
6. then you should be able to `from remedian.remedian import Remedian`

# Usage

See the [example in the docs](https://remedian.readthedocs.io/en/latest/examples.html).

# References

> P.J. Rousseeuw, G.W. Bassett Jr., "The remedian: A robust averaging method
  for large data sets", Journal of the American Statistical Association, vol.
  85 (1990), pp. 97-104

> M. Chao, G. Lin, "The asymptotic distributions of the remedians", Journal of
  Statistical Planning and Inference, vol. 37 (1993), pp. 1-11

> Domenico Cantone, Micha Hofri, "Further analysis of the remedian algorithm",
  Theoretical Computer Science, vol. 495 (2013), pp. 1-16


docs/changelog.rst
:orphan:

.. _whats_new:


Changelog
=========

Here we list what's new in Remedian.

.. _current:

Current
-------

Changelog
~~~~~~~~~

- New example added by `Stefan Appelhoff`_ "examples/pros_cons_remedian.py"

Bug
~~~

- no entries yet

API
~~~

- no entries yet

.. _v0.1:

|
|
|

v0.1
----

Changelog
~~~~~~~~~

- Initial release

Authors who contributed to this release
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- `Stefan Appelhoff`_

.. _Stefan Appelhoff: http://stefanappelhoff.com/


test/test_remedian.py
"""Tests for the Remedian class."""
import pytest
import numpy as np

from remedian.remedian import Remedian


def test_wrong_input():
    """Test behavior if wrong input is given at initialization."""
    # No input
    try:
        Remedian()
    except TypeError:
        pass

    # obs_size is not a shape (tuple)
    try:
        Remedian(1, 2, 3)
    except TypeError:
        pass

    # Normal call of Remedian without error
    assert Remedian((1,), 2, 3)

    # Observations <= 1 are impossible
    for n_obs in range(-1, 2):
        try:
            print(n_obs)
            Remedian((1,), n_obs, 3)
        except ValueError:
            pass

    # Negative number of input arrays are impossible
    try:
        Remedian((1,), 2, -3)
    except ValueError:
        pass


def test_initial_values():
    """Test the initial values."""
    random_shape = np.random.randint(1, 100, np.random.randint(1, 5, 1)[0])
    obs_size = np.random.random((random_shape)).shape
    n_obs = np.random.randint(1, 100, 1)[0]
    t = np.random.randint(1, 100, 1)[0]
    r = Remedian(obs_size, n_obs, t)
    assert r.obs_size == list(obs_size)
    assert r.n_obs == n_obs
    assert r.t == t


def test_calc_remedian():
    """Test whether the calculation of a remedian works."""
    random_shape = np.random.randint(1, 3, np.random.randint(1, 3, 1)[0])
    obs_size = np.random.random((random_shape)).shape
    n_obs = np.random.randint(1, 50, 1)[0]
    t = np.random.randint(1, 50, 1)[0]
    r = Remedian(obs_size, n_obs, t)
    assert r.remedian is None
    for i in range(t):
        obs = np.random.random(obs_size)
        r.add_obs(obs)
    assert isinstance(r.remedian, np.ndarray)
    assert r.remedian.shape == obs_size


def test_too_many_obs():
    """Test behavior if more obs are given than t specifies."""
    obs_size = (2, 5)
    n_obs = 3
    t = 10
    r = Remedian(obs_size, n_obs, t)
    for i in range(t):
        obs = np.random.random(obs_size)
        r.add_obs(obs)

    # One more should be one too much
    with pytest.raises(RuntimeError):
        r.add_obs(np.random.random(obs_size))


def test_wrong_obs_size():
    """Test behavior if obs_size is unexpected at call to add_obs."""
    random_shape = np.random.randint(1, 3, np.random.randint(1, 3, 1)[0])
    obs_size = np.random.random((random_shape)).shape
    n_obs = np.random.randint(1, 50, 1)[0]
    t = np.random.randint(1, 50, 1)[0]
    r = Remedian(obs_size, n_obs, t)

    # We give a wrong obs_size
    wrong_obs_size = tuple(np.asarray(obs_size)+1)
    with pytest.raises(ValueError):
        r.add_obs(np.random.random(wrong_obs_size))


def test_true_median():
    """Test that Remedian==median if n_obs==t or n_obs>t."""
    obs_size = (2, 5)
    t = 10

    # Test for n_obs == t
    n_obs = t
    r = Remedian(obs_size, n_obs, t)

    res = []
    for i in range(t):
        obs = np.random.random(obs_size)
        r.add_obs(obs)
        res.append(obs)

    true_median = np.median(np.asarray(res).squeeze(), axis=0)
    remedian_median = r.remedian.squeeze()
    assert true_median.shape == remedian_median.shape
    np.testing.assert_array_equal(true_median, remedian_median)

    # Now test for n_obs > t
    n_obs = t + np.random.randint(1, 5, 1)[0]
    r = Remedian(obs_size, n_obs, t)

    res = []
    for i in range(t):
        obs = np.random.random(obs_size)
        r.add_obs(obs)
        res.append(obs)

    true_median = np.median(np.asarray(res).squeeze(), axis=0)
    remedian_median = r.remedian.squeeze()
    assert true_median.shape == remedian_median.shape
    np.testing.assert_array_equal(true_median, remedian_median)


remedian/tests/test_remedian.py
"""Tests for the Remedian class."""
import pytest
import numpy as np

from remedian.remedian import Remedian


def test_wrong_input():
    """Test behavior if wrong input is given at initialization."""
    # No input
    with pytest.raises(TypeError):
        Remedian()

    # obs_size is not a shape (tuple)
    with pytest.raises(TypeError):
        Remedian(1, 2, 3)

    # Normal call of Remedian without error
    assert Remedian((1,), 2, 3)

    # Observations <= 1 are impossible
    with pytest.raises(ValueError):
        Remedian((1,), 0, 3)

    # Negative number of input arrays are impossible
    with pytest.raises(ValueError):
        Remedian((1,), 2, -3)


def test_initial_values():
    """Test the initial values."""
    random_shape = np.random.randint(1, 100, np.random.randint(1, 5, 1)[0])
    obs_size = np.random.random((random_shape)).shape
    n_obs = np.random.randint(2, 100, 1)[0]
    t = np.random.randint(1, 100, 1)[0]
    r = Remedian(obs_size, n_obs, t)
    assert r.obs_size == list(obs_size)
    assert r.n_obs == n_obs
    assert r.t == t


def test_calc_remedian():
    """Test whether the calculation of a remedian works."""
    random_shape = np.random.randint(1, 3, np.random.randint(1, 3, 1)[0])
    obs_size = np.random.random((random_shape)).shape
    n_obs = np.random.randint(1, 50, 1)[0]
    t = np.random.randint(1, 50, 1)[0]
    r = Remedian(obs_size, n_obs, t)
    assert r.remedian is None
    for i in range(t):
        obs = np.random.random(obs_size)
        r.add_obs(obs)
    assert isinstance(r.remedian, np.ndarray)
    assert r.remedian.shape == obs_size


def test_too_many_obs():
    """Test behavior if more obs are given than t specifies."""
    obs_size = (2, 5)
    n_obs = 3
    t = 10
    r = Remedian(obs_size, n_obs, t)
    for i in range(t):
        obs = np.random.random(obs_size)
        r.add_obs(obs)

    # One more should be one too much
    with pytest.raises(RuntimeError):
        r.add_obs(np.random.random(obs_size))


def test_wrong_obs_size():
    """Test behavior if obs_size is unexpected at call to add_obs."""
    random_shape = np.random.randint(1, 3, np.random.randint(1, 3, 1)[0])
    obs_size = np.random.random((random_shape)).shape
    n_obs = np.random.randint(1, 50, 1)[0]
    t = np.random.randint(1, 50, 1)[0]
    r = Remedian(obs_size, n_obs, t)

    # We give a wrong obs_size
    wrong_obs_size = tuple(np.asarray(obs_size)+1)
    with pytest.raises(ValueError):
        r.add_obs(np.random.random(wrong_obs_size))


def test_true_median():
    """Test that Remedian==median if n_obs==t or n_obs>t."""
    obs_size = (2, 5)
    t = 10

    # Test for n_obs == t
    n_obs = t
    r = Remedian(obs_size, n_obs, t)

    res = []
    for i in range(t):
        obs = np.random.random(obs_size)
        r.add_obs(obs)
        res.append(obs)

    true_median = np.median(np.asarray(res).squeeze(), axis=0)
    remedian_median = r.remedian.squeeze()
    assert true_median.shape == remedian_median.shape
    np.testing.assert_array_equal(true_median, remedian_median)

    # Now test for n_obs > t
    n_obs = t + np.random.randint(1, 5, 1)[0]
    r = Remedian(obs_size, n_obs, t)

    res = []
    for i in range(t):
        obs = np.random.random(obs_size)
        r.add_obs(obs)
        res.append(obs)

    true_median = np.median(np.asarray(res).squeeze(), axis=0)
    remedian_median = r.remedian.squeeze()
    assert true_median.shape == remedian_median.shape
    np.testing.assert_array_equal(true_median, remedian_median)


examples/use_remedian.py
"""
============
Use Remedian
============

This example exists to demonstrate the basic usage of
:class:`remedian.Remedian`.

"""

# Authors: Stefan Appelhoff <stefan.appelhoff@mailbox.org>
# License: MIT

###############################################################################
# First we import what we need for this example.
import matplotlib.pyplot as plt
import numpy as np
from remedian import Remedian

###############################################################################
# Now we make up some data SHAPE reflecting the shape of the data that we
# want to calculate the Remedian on.
# We can have data of any shape ... e.g., 3D:
data_shape = (2, 3, 4)

###############################################################################
# Now we have to decide how many data observations we want to load into
# memory at a time before computing a first intermediate median from it
n_obs = 100

###############################################################################
# Pick some example number ... assume we have ``t`` arrays of shape
# ``data_shape`` that we want to summarize with Remedian
t = 500

###############################################################################
# Initialize the object
r = Remedian(data_shape, n_obs, t)

###############################################################################
# Feed it the data ... for now, we just generate the data randomly on the go
# ... also save the actual data for comparison with true median
my_data = []
for obs_i in range(t):
    obs = np.random.random(data_shape)  # We just generate some random data
    r.add_obs(obs)
    my_data.append(obs)  # save random data for later (for example purposes)

###############################################################################
# Now we have the Remedian in ``r.remedian``
# Let's summarize the results
x = np.median(np.asarray(my_data).squeeze(), axis=0)  # the *true* median
y = r.remedian  # the *remedian*
xydiff = x-y


###############################################################################
# Now let's make a plot! We force the 3D data to 2D using
# :func:`numpy.reshape`.

# For colorbar scaling
vmax = np.max([np.abs(x), np.abs(y), np.abs(xydiff)])
vmin = -vmax


fig, (ax1, ax2, ax3) = plt.subplots(1, 3)

for data_to_plot, ax, name in zip([x, y, xydiff], [ax1, ax2, ax3],
                                  ['True median', 'Remedian', 'Difference']):
    im = ax.imshow(data_to_plot.reshape(1, -1), aspect='auto', cmap='bwr',
                   vmin=vmin, vmax=vmax)
    ax.set_xticks([])
    ax.set_yticks([])
    ax.set_title(name)

cbar = plt.colorbar(im, ax=ax)


examples/pros_cons_remedian.py
"""
=========================
Pros and Cons of Remedian
=========================

In this example we test :class:`remedian.Remedian` against :func:`numpy.median`
in terms of three parameters:

- computation time
- mean squared error compared to true median
- memory needed for computation

"""

# Authors: Stefan Appelhoff <stefan.appelhoff@mailbox.org>
# License: MIT

###############################################################################
# First we import what we need for this example.
from timeit import default_timer as timer

import matplotlib.pyplot as plt
import numpy as np
from remedian import Remedian

###############################################################################
# We start by generating two datasets that we want to compute the median of
# along the last dimension (or approximate the median via Remedian).

n = 500  # this is the dimension across which we want to compute the median
data_shape = (50, 50, n)
data_unif = np.random.random(data_shape)
data_norm = np.random.randn(*data_shape)

###############################################################################
# Now let's measure three parameters: The time it takes to compute the
# Remedian, the memory it needs to keep the data loaded, and the mean
# squared error compared to the *true* median.
#
# We will do this for different Remedian ``n_obs`` parameters. ``n_obs``
# determines how many intermediate medians are taken to approximate the median.
# Low ``n_obs`` means little memory required, but also a more noisy estimation
# of the median. High ``n_obs`` or ``n_obs`` equal to ``n`` equals the *true*
# median.

datas = [data_unif, data_norm]
n_obses = range(2, n+1)  # n_obs cannot be 1 or smaller

# The three parameters we want to collect data on
memory_needed = np.zeros((len(datas), len(n_obses)))
compute_times = np.zeros((len(datas), len(n_obses)))
mses = np.zeros((len(datas), len(n_obses)))

# Start collecting the data across our two generated datasets
for idata, data in enumerate(datas):

    # calculate true median for this dataset
    median = np.median(data, axis=-1)

    # Calculate the Remedian for this dataset, given differen ``n_obs`` params
    for in_obs, n_obs in enumerate(n_obses):

        start = timer()
        # New Remedian object
        rem = Remedian(data_shape[:-1], n_obs, n)

        # calculate the remedian
        for data_idx in range(n):
            rem.add_obs(data[..., data_idx])

        approx_median = rem.remedian
        end = timer()

        # Time elapsed in seconds
        compute_times[idata, in_obs] = end - start

        # Memory needed in bytes
        memory_needed[idata, in_obs] = n_obs * (data[..., data_idx].size *
                                                data[..., data_idx].itemsize)

        # mean square error from true median
        mses[idata, in_obs] = np.mean((approx_median-median)**2)

###############################################################################
# We ran our data collection for ``n_obs`` parameters up to ``n``, which
# equals the median as given by :func:`np.median`. Let's plot the results.

fig, axs = plt.subplots(3, 1, figsize=(10, 7.5), sharex=True)

names = ['memory_needed_[bytes]', 'computation_time_[s]', 'mean_squared_error']
results = [memory_needed, compute_times, mses]
for ax, name, result in zip(axs, names, results):

    ax.plot(n_obses, result[0, ...], label='data_unif')
    ax.plot(n_obses, result[1, ...], label='data_norm')
    ax.axvline(n_obses[-1], color='black', linestyle='--',
               label='*true* median')

    ax.set_ylabel(name)
    if name.startswith('memory'):
        ax.legend()

ax.set_xlabel('`n_obs` parameter')
fig.tight_layout()

###############################################################################
# In summary:
#
# - Remedian needs less memory than the *true* median
# - Remedian's ``n_obs`` parameter can be set in a "good" or "bad" way
#   regarding the mean squared error compared to the *true* median
# - Remedian's timing seems to be relatively unaffected by the ``n_obs``
#   parameter, and close to the timing of the *true* median.


remedian/remedian.py
# -*- coding: utf-8 -*-
"""Contains an implementation of Remedian."""

# Author: Stefan Appelhoff <stefan.appelhoff@mailbox.org>
# License: MIT

import numpy as np


class Remedian():
    """Remedian object for a robust averaging method for large data sets.

    Implementation of the Remedian algorithm, see [1]_ [2]_ [3]_ for
    references. This algorithm is used to approximate the median of several
    data chunks if these data chunks cannot (or should not) be loaded into
    memory at once. See "Notes" section for further information.

    Parameters
    ----------
    obs_size : ndarray
        The shape of each data chunk (=observation) to be fed into the Remedian
        object.
    n_obs : int
        The number of observations to be stored within each array.
        If `n_obs` >= `t`, Remedian will equal the median. The smaller this
        parameter, the fewer data have to be loaded into memory at once, but
        the less accurate the approximation of the median will be.
    t : int
        The total number of observations from which a median should be
        approximated.

    Attributes
    ----------
    obs_count : int
        Counter of number of observations that have already been given
        to the Remedian object.
    remedian : None | ndarray, shape(obs_size)
        The calculated remedian of the same shape as the input data.
        Will be None until all observations `n_obs` have been fed into
        the object using the add_obs method.

    Notes
    -----
    Given a data chunk of size `obs_size`, and `t` data chunks overall, the
    Remedian class sets up a number `k_arrs` of arrays of length `n_obs`.

    The median of the `t` data chunks of size `obs_size` is then approximated
    as follows: One data chunk after another is fed into the `n_obs` positions
    of the first array. When the first array is full, its median is calculated
    and stored in the first position of the second array. After this, the first
    array is re-used to fill the second position of the second array, etc.
    When the second array is full, the median of its values is stored in the
    first position of the third array, and so on.

    The final "Remedian" is the median of the last array, after all `t` data
    chunks have been fed into the object.

    In other words, given an n-dimensional array, the Remedian class
    approximates the median of this array across the ith dimension and you have
    to break up your n-dimensional array into `t` n-1-dimensional arrays that
    are given to Remedian one after another.

    References
    ----------
    .. [1] P.J. Rousseeuw, G.W. Bassett Jr., "The remedian: A robust averaging
       method for large data sets", Journal of the American Statistical
       Association, vol. 85 (1990), pp. 97-104

    .. [2] M. Chao, G. Lin, "The asymptotic distributions of the remedians",
       Journal of Statistical Planning and Inference, vol. 37 (1993), pp. 1-11

    .. [3] Domenico Cantone, Micha Hofri, "Further analysis of the remedian
       algorithm", Theoretical Computer Science, vol. 495 (2013), pp. 1-16

    """

    def __init__(self, obs_size, n_obs, t):
        """Initialize the Remedian object.

        See class docstring for more thorough information.

        Parameters
        ----------
        obs_size : ndarray
            Size of the observations. Must be (1,) for scalars.
        n_obs : int
            Observations per array.
        t : int
            Number of total observations.

        """
        if n_obs <= 1:
            raise ValueError('`n_obs` of <= 1 does not make sense.')

        self.obs_size = list(obs_size)
        self.n_obs = n_obs
        self.t = t

        # Calculate the number of arrays needed and their sizes
        self.k_arrs = self._calc_k_arrs()
        self.k_arr_sizes = self._calc_k_arr_sizes()

        # Initialize the arrays
        self.arrs = [np.zeros(self.obs_size+[s]) for s in self.k_arr_sizes]

        # counter for observations within each array
        self.obs_idx_counter = [0 for arr in range(self.k_arrs)]

        # Modulos of observations to assign to correct array later
        self.modulos = [self.n_obs**i for i in range(1, 1+self.k_arrs)]

        # Counter of received observations
        self.obs_count = 0

        # Set the median value to None until we have it
        self.remedian = None

    def _calc_k_arrs(self):
        """Calculate number of arrays to accommodate the observations."""
        tmp = self.n_obs
        k_arrs = 1
        while tmp <= self.t:
            tmp *= self.n_obs
            k_arrs += 1
        return k_arrs

    def _calc_k_arr_sizes(self):
        """Calculate the size of each array to accomodate the observations."""
        k_arr_sizes = [self.n_obs for i in range(self.k_arrs)]
        k_arr_sizes[-1] = int(np.ceil(self.t / (self.n_obs**(self.k_arrs-1))))
        return k_arr_sizes

    def add_obs(self, obs):
        """Add an observation to the Remedian.

        Parameters
        ----------
        obs : ndarray, shape(obs_size)
            A single data observation.

        """
        # We only work if:
        # ... we get an observation of correct size
        # ... we have not yet received all observations already
        if list(obs.shape) != self.obs_size:
            raise ValueError('Expected observation of size {} but received: '
                             '{}'.format(self.obs_size, list(obs.shape)))
        if self.obs_count > (self.t - 1):
            raise RuntimeError('Already collected {} observations out of t={} '
                               'The remedian is {}'.format(self.obs_count,
                                                           self.t,
                                                           self.remedian))

        # We accept a new observation
        self.obs_count += 1

        # Add the data to the first array
        # and increment the counter for the next data
        obs_idx = self.obs_idx_counter[0]
        self.arrs[0][..., obs_idx] = obs
        self.obs_idx_counter[0] += 1

        # We can notice whenever an array is full using modulo operations
        # on the observation counter self.obs_count.
        # When an array is full, calculate the median and put the result
        # into the next array. Then reset the counters and start filling
        # previous arrays again
        for arr_i, mod in enumerate(self.modulos):
            if self.obs_count % mod == 0:
                data = self.arrs[arr_i]
                m_tmp = np.median(data, axis=-1, overwrite_input=True)
                self.arrs[arr_i+1][..., self.obs_idx_counter[arr_i+1]] = m_tmp
                self.obs_idx_counter[arr_i+1] += 1
                self.obs_idx_counter[arr_i] = 0

        # If all observations have been received,
        # calculate the median of the last array.
        # This is the robust approximation of the median
        if self.obs_count == self.t:
            self.remedian = np.median(self.arrs[-1], axis=-1,
                                      overwrite_input=True)
